Rework scene preview thumbnails (reverted)

daniel080400 opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

Supersedes #102263
Will need to disable code from #96544 because of conflict

Make scene files (.glb, gltf, .tscn, .scn) to always generate missing/outdated thumbnails, captures in a fixed angle, and fits all visual elements into the image, operation is done on a separate thread. Respects environment settings at ```rendering/environment/defaults/default_clear_color```.

This PR also fixes preview image caching, when testing, **file has to be changed to generate a new thumbnail.**  (md5 != last_md5) 


# Caveats
1. Loses the ability to capture custom thumbnails with view. Open and save a scene still result in fixed view thumbnails, but as discussed in #102263, I think it would be better to implement custom thumbnails in another better approach.

2. If scene has particle with large bounding box, preview camera will be set far away because of it. Meshes might be small in view. This is done intentionally, to accommodate scenes that are basically particle effects.
![image](https://github.com/user-attachments/assets/b46ed441-28a9-4031-b43b-926a274c56b0)

# 3D
Camera will always point at world center, dolly out at 3/4 view angle, until all contents fit into screen.
For particles, fast-forwards them it to (lifetime * 0.5)  in order to render something, with fixed seed.

Practical:
![image](https://github.com/user-attachments/assets/13f8d3ee-31a5-46f6-9acc-673fdd8f48e0)

Production:
![image](https://github.com/user-attachments/assets/a4b2b10c-897e-49ab-8b78-c39931efc521)

# 2D
Render in two passes, 2D and GUI, then combines 2D + GUI captures for the final thumbnail image. 

Does **NOT** account the world center for 2D capture, this decision is made because the use case that will benefit most from scene thumbnails are prefabs, which normally don't care about world coordinates, thus dedicating more pixels to actual contents.

![image](https://github.com/user-attachments/assets/9912bd54-6361-4fc9-9832-8a220b0f5cb3)

TODO:
 - [x] Generate scene thumbnails on file create/rename/save/duplicate/import
 - [x] Only apply preview lighting to scenes that have 0 light nodes
 - [x] Prevent tool scripts from running when generating thumbnails
 - [x] Support for 2d scenes (Scenes with 0 Node3D count)
 - [x] Make thumbnail creation thread-safe
 - [x] Do proper cache validation at ```editor_resource_preview.cpp``` to prevent outdated thumbnails

Test projects:
[rework_scene_thumbnails_test_proj.zip](https://github.com/user-attachments/files/19721611/rework_scene_thumbnails_test_proj.zip)

[junk_files.zip](https://github.com/user-attachments/files/20516654/junk_files.zip)

<details>
<summary>DEV NOTES</summary>

 - Rip off scripts by using ```SceneState::get_bundled_scene()``` and ```SceneState::set_bundled_scene()``` to change the data of PackedScene before instantiating it.
 - In ```editor_resource_preview.cpp```, suspect cache validation is somewhat wrong within these lines, will return early with a outdated thumbnail image:
    - line 276-281
    - line 305-307
    - line 488-491

 - Need to disable the old ```EditorNode::_save_scene_with_preview``` for this PR
 - Will result in outdated thumbnail if user quickly (save->edit->save) a scene within ~0.5 seconds, this should not happen, is there something wrong with how ```EditorResourcePreview::queue``` is managed?
 - Control anchors do not play well with Camera2D, thus 2d scene thumbnail need to be rendered twice (gui + node2d)
 - It might be possible to draw viewport texture outside of the tree using ```RenderingServer::viewport_create()``` with ```EditorResourcePreviewGenerator::draw_requester.request_and_wait()```, also being thread safe.

</details>
___

- *Production edit: This closes https://github.com/godotengine/godot-proposals/issues/7232.*

ryevdokimov commented:
>The top one problem to solve for this PR to work is to find a way to instantiate scenes without triggering tool scripts

I'm sure there is a valid reason for this, but I'm curious since I haven't given it much thought, but what is the issue with tool scripts?

The reason I ask is because I have a project that uses tools scripts to automatically and dynamically generate objects to build out a scene when it is first opened, so if the tool scripts were not ran it would be an empty or broken scene.

daniel080400 commented:
> I'm sure there is a valid reason for this, but I'm curious since I haven't given it much thought, but what is the issue with tool scripts?

Mainly about unexpected behavior, running scripts even when scenes are not opened might confuse many, like if someone decided to print() messages or access filesystem in their tool scripts. As for stability concerns, I think if a tool script will crash the editor, it might be written in a very wrong way at the first place.

ryevdokimov commented:
I think in general if a user is using tool scripts it is at their own risk, and they should handle edge cases themselves, but new functionality should still accommodate their existence because they're vital for various workflows, users just need the tools to be able to minimize that risk. 

This isn't to say that there is a solution to this problem, and we may just have to disable tool scripts while generating thumbnails and deal with it, but just food for thought.

daniel080400 commented:
Also in this PR, thumbnails are generated in a separate thread, and works by adding the preview scene under the current **EditorNode** to make it render (in a SubViewport), this means if your tool script is not thread-safe (by calling **add_child()** in **_ready()** ), it will cause errors.

I experimented if creating a separate SceneTree is possible, turns out it's not easy, so might as well just focus on providing this feature in a safest way.

JoNax97 commented:
This is looking very good! 
I think that, if merged, it would be enough to close https://github.com/godotengine/godot-proposals/issues/7232. Additional tweaks and personalization settings can be proposed by users after real world usage. 

I'm very glad someone got to implementing this!

daniel080400 commented:
I discovered there's no way to accurately determine the global Rect2 of each Node2D (some may have textures, or meshes that have various sizes), I decided to only address the most important types of Node2D into Rect2 calculation.

Fow now only Sprite2D is accounted into thumbnails, plan to implement for AnimatedSprite2D, MeshInstance2D, MultimeshInstance2D, TileMapLayer, Polygon2D, TouchScreenButton later. By then will consider 2D scene thumbnail as completed.

![Screenshot from 2025-02-06 00-12-59](https://github.com/user-attachments/assets/c94eb246-d38f-4373-8e2a-b3580ff99236)

daniel080400 commented:
2D scene thumbnail is completed, Node2D types that participates scene Rect2 calculations are:
**Sprite2D, AnimatedSprite2D, MeshInstance2D, MultimeshInstance2D, TileMapLayer, Polygon2D, Line2D, TouchScreenButton**

![image](https://github.com/user-attachments/assets/9912bd54-6361-4fc9-9832-8a220b0f5cb3)

daniel080400 commented:
All planned features completed. Squashed and green lighting this. Now for testing and clean up.

Calinou commented:
> consider web editor could only handle 2K textures at most in rendering, correct me if wrong

The universal baseline across all platforms Godot 4 runs on is 4096x4096 nowadays.

daniel080400 commented:
> The universal baseline across all platforms Godot 4 runs on is 4096x4096 nowadays.

Thanks to clarify that, anyway this should not be a problem now. I updated to only zoom out the preview Camera2D when capturing 2d thumbnails (previously is done by scaling the SubViewport size).

Calinou reviewed (commented):
Tested locally (rebased on top of `master` https://github.com/godotengine/godot/commit/39c201ca58a8ddf1e99e62c8453294a2662a2fd8), I get an issue in 2D. The [2D platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) freezes when opened in the editor after the "restoring scene layout" popup is done running. This seems to occur with every 2D project I've tried.

In comparison, it works in a 3D project like [Truck Town](https://github.com/godotengine/godot-demo-projects/tree/master/2d/truck_town) and [Platformer](https://github.com/godotengine/godot-demo-projects/tree/master/3d/platformer):

![image](https://github.com/user-attachments/assets/b7937f87-99e2-4ecc-aeb0-3970ecbd4374)

Note that `player.tscn` is very small in the 3D platformer demo because the Decal node's AABB is taken into account. (The decal is used for a blob shadow.) Decal nodes' AABBs should probably be excluded if possible, as we only care about the surface the decal is being projected on for the preview, not the decal itself.

This PR also gets rid of the progress dialog that appears when saving a scene:

![image](https://github.com/user-attachments/assets/763a8332-6041-4fc5-a6bd-b198ca3bf6dd)

I consider this to be a good change for scenes that save in less than 1 second – this should also help resolve https://github.com/godotengine/godot/issues/78753. However, for scenes that take a while to save, we should have some kind of progress dialog back (or at least a visible notice somewhere in the editor). You can create a scene that takes a while to save by adding a VoxelGI node, setting it to 512 subdivisions, baking it and choosing **Make Unique** on the VoxelGI node's **Data** property.

Additionally, you could do the same as https://github.com/godotengine/godot/pull/97468 to improve preview quality using MSAA and SSAA together. This should only increase the GPU requirements minimally since the previews are low-resolution. To cater for 2D previews, also enable 2D MSAA when relevant, not just 3D MSAA.

For a future PR, we can also further improve the preview by adding a checkered floor below the object and enabling shadows on the DirectionalLight3D. This way, you get a sense of the object's scale.

daniel080400 commented:
>I get an issue in 2D. The [2D platformer demo](https://github.com/godotengine/godot-demo-projects/tree/master/2d/platformer) freezes when opened in the editor after the "restoring scene layout" popup is done running. This seems to occur with every 2D project I've tried.

Issue caused by renderer being set to **Compatibility**, ```EditorResourcePreview``` will force all preview plugins to run on the main thread. I tweaked the code to have it run safely on single thread. But raises another issue, on project startup, now scans for the whole filesystem and generate missing thumbnails all at once, it would cause clutter if the project is too big. I'm experimenting on why we forced single-threaded on preview plugins when using Compatibility renderer, and look for a workaround to enable multi-thread if possible.

Calinou reviewed (commented):
The 2D platformer demo no longer freezes on opening. It still got stuck for a while while generating previews, but I couldn't reproduce this a second time, even after removing `$HOME/.cache/godot/` and `.godot/` in the project folder.

Also, messages about 2D MSAA not being supported in GLES3 are spammed when the editor is open. To fix this, wrap the call that sets 2D MSAA with this:

```cpp
if (RS::get_singleton()->get_current_rendering_method_name() != "gl_compatibility") {
    ...
}
```

Lastly, the default canvas texture filter project setting is not applied when generating previews, which means sprites always use linear filtering. This can cause them to look blurry in the preview:

![image](https://github.com/user-attachments/assets/c5f33db8-d886-4d64-90c5-2ddb6e409a5c)

To fix this, set the default canvas texture filter and repeat settings on the SubViewport used for thumbnail rendering according to the project settings `rendering/textures/canvas_textures/default_texture_filter` and `rendering/textures/canvas_textures/default_texture_repeat`.

Calinou commented on editor/editor_node.cpp:
```diff
@@ -1726,7 +1726,9 @@ void EditorNode::_find_node_types(Node *p_node, int &count_2d, int &count_3d) {
 
+// **Deprecated** now generates preview at EditorPackedScenePreviewPlugin
 void EditorNode::_save_scene_with_preview(String p_file, int p_idx) {
+	return; // Disable
```
This method is still indirectly called by `EditorInterface.save_scene_as()` (when the second parameter is `true` which is the default). This is exposed to scripting.

However, can't we remove this particular method and stop calling it in `save_scene_to_path()`?

The `with_preview` parameter of `EditorInterface.save_scene_as()` should then be marked as deprecated in the documentation:

```xml
		<method name="save_scene_as">
			<return type="void" />
			<param index="0" name="path" type="String" />
			<param index="1" name="with_preview" type="bool" default="true" />
			<description>
				Saves the currently active scene as a file at [param path].
				[b]Note:[/b] The [param with_preview] parameter has no effect.
			</description>
		</method>
```

daniel080400 commented:
Changes made:
1. Don't use 2D MSAA in scene previews when renderer is set to Compatibility
2. Respects project settings ```rendering/textures/canvas_textures/default_texture_filter``` and ```rendering/textures/canvas_textures/default_texture_repeat``` when generating 2d previews.
3. Clean up resources properly when generating preview for scene that contains GridMap.
4. Remove ```EditorNode::_save_scene_with_preview```, route ```EditorInterface.save_scene_as()``` to always call ```EditorNode::_save_scene```, also updates comments and document.

daniel080400 commented:
damn, accidentally merged the whole change from master, let me revert back

daniel080400 commented:
Good again. Sorry for the disturbance.

JoNax97 commented:
Do you think it will be possible to merge this in time for 4.5?

Mickeon commented:
Would need a rebase, first of all

daniel080400 commented:
Changes made:
1. Rebase to master commit 215acd52e82f4c575abb715e25e54558deeef998
2. Account for nested tool scripts (scene with tool scripts inside a scene) while generating thumbnails:
 ```
bool EditorPackedScenePreviewPlugin::_remove_scripts_from_packed_scene(Ref<PackedScene> pack) const {
		...

		if (Object::cast_to<PackedScene>(edited_variants[i])) {
			_remove_scripts_from_packed_scene(edited_variants[i]); // Account for nested tool scripts
		}
 ```

fire commented:
Do you have a test case project for reviewing?

daniel080400 commented:
Changes made:
1. Fixed various bug since rebased to master at 215acd52e82f4c575abb715e25e54558deeef998
2. Fixed bug in ```EditorMaterialPreviewPlugin``` where material resources will be assigned to a incorrect thumbnail while using Forward+ or Mobile renderer. (Basically shares the same fix in this PR)

@fire Test case project:
[rework_scene_thumbnails_test_proj.zip](https://github.com/user-attachments/files/19721611/rework_scene_thumbnails_test_proj.zip)
Also updated in main comment.

KoBeWi commented:
Tested with my project, I get spammed with errors:
![image](https://github.com/user-attachments/assets/73255617-143d-42f9-95a9-af9f863c5bf7)
Looks related to custom signals.

Also most of my scenes ended up with black previews for some reason:
![image](https://github.com/user-attachments/assets/71808513-46c8-4bb4-8522-9c58c446ac8f)
No idea what could be causing it. Level scenes all have similar structure and size, but not all of them are black. Non-level scenes (enemies, objects etc.) all have black previews.

I also have a 2D scene with 3D object. Only the 3D object is rendered in the preview, the 2D part is completely skipped (even though the scene is 2D and the object is shown in SubViewport).

daniel080400 commented:
@KoBeWi

> Tested with my project, I get spammed with errors:
> Looks related to custom signals.

Confirmed to only happen with custom signals that are connected via the "Node" dock, I'll look into this.

> Also most of my scenes ended up with black previews for some reason:
> No idea what could be causing it. Level scenes all have similar structure and size, but not all of them are black. Non-level > scenes (enemies, objects etc.) all have black previews.

It'll help me a lot if you can provide further information:
1. Do you have a 3D node somewhere in those scenes?
2. Do you set your default environment's color to black?
3. Is there any node in those scenes that is able to render the whole scene black when visible?

> I also have a 2D scene with 3D object. Only the 3D object is rendered in the preview, the 2D part is completely skipped.

This is done intentionally, in this PR we follow this sequence of logic:
1. Count the scene's Node3D, if is greater than 1, render a 3D preview.
2. Count the scene's (Node2D + Control) node, if is greater than 1, render a 2D preview. 
3. Return an empty image if no conditions above are met.

As I initially thought there's no reason to have 3D nodes inside a 2D scene, if it does, it should be 3D scene with 2D nodes presenting as UI. But after looking at your post I re-think it, the quickest fix now is ignore anything under a SubViewport node and don't accumulate its children to 3d, 2d node counts. So if a 2D scene has SubViewport rendering 3D scene it will still be considered a 2D scene and render a 2D preview.

Secondly, I don't merge 3D + 2D previews for a reason, when have a 3D scene with a 2D scene placed inside used as a UI, the preview image will be filled up by 2D preview even if it's a 3D scene.

Thank you for your time, appreciated.

KoBeWi commented:
> Do you have a 3D node somewhere in those scenes?

No.

> Do you set your default environment's color to black?

Yes.

> Is there any node in those scenes that is able to render the whole scene black when visible?

No.

As I said, the level scenes are very similar. A Node2D root with TileMapLayers and some entities. Other scenes are mostly CharacterBody2D with sprite.

I could try copying them to another project and see if the problem remains.

daniel080400 commented:
@KoBeWi 
> I could try copying them to another project and see if the problem remains.

I suspect it'll remain the same, if you have time, can you copy one .tscn that has black previews, try deleting nodes inside it and re-save the file each time until the preview image shows up correctly? It might help us find out what specific node is causing the problem and narrow down the scope.

I'd tested a scene with multiple TileMapLayers and it's fine on my side.

KoBeWi commented:
No idea what was that bug, maybe it was caused by old previews (generated before this PR).
The reason my scenes stayed black was that I tried to recreate the thumbnail without any changes. After I simply renamed a node and saved scene, the thumbnail was generated correctly.
I then proceeded to delete thumbnail cache folder and restart the project. Everything it being generated properly.

Sounds like the old cache should be invalidated somehow. This PR will have no effect if people have old previews (they are stored in editor data folder, so deleting .godot won't help).

daniel080400 commented:
> Sounds like the old cache should be invalidated somehow.

Okay I'll do versioning for cache files (store a version=2 property), for now it just validates cache with file md5 checksum so that was expected to happen on old projects.

daniel080400 commented:
Changes made in commit c8614c6:
1. Increment preview thumbnail metadata version to **2**, now old projects loaded after this PR will re-generate all thumbnails.
2. Fix error when generating previews of a scene that has custom signals connected via "Node" dock.
3. When counting 2d/3d nodes in a scene to determine whether to generate a 2d/3d preview, do not count nodes under a ```SubViewport``` node. In the other hand, if the root node is set to ```SubViewport```, its children will still be counted.

---

> I think it might be clearer for users to simply render 2D or 3D based on the type of the root node

@JoNax97 That can be discussed. The behavior I use now is inherited from the old way how we captured thumbnails from current viewport view, it determines the preview type based on 2d 3d node counts, not the root node, and I don't know if it will confuse the users if that behavior has changed. Like some users might use a "WorldEnvironment" as root like a monster, and still expects the thumbnail to generate its 2d/3d nodes' preview.

@KoBeWi Bugs you encountered yesterday should all be fixed now.

To make previews render correctly, there's some hacky code such as waiting frames from ```RenderingServer```, requesting viewports to render multiple times involved in this PR. I have stress-tested the code (eg: run the game while previews are still generating) and never encountered any bugs or crashes. But if someone tested out some instability, I can move the whole process back to single-threaded, or stop calling ```RenderingServer``` directly (Which is already done in this way with Mesh, Material previews, and I see no issue reports) and use a workaround.

daniel080400 commented:
Tested with a larger project (1GB in total, 100+ glb and .tscn files, lots of 2K textures)

All thumbnails are created correctly, but have some issues:
1. When using Forward/ Mobile renderer (Multi threaded thumbnail creation), the editor stays responsive, but the viewport flickers (to a point it's irritating) while generating thumbnails
2. When using Compatibility renderer (Single threaded thumbnail creation), the editor freezes until all thumbnails in the project are generated (waited for 2 minutes in my project)

I'll investigate issue 1 tomorrow, I feel like it's not a hard to fix issue.

For issue 2 I have no idea how to get around it, since it's a renderer limit. Some "solutions" in my mind:
* Spawn a progress bar saying "Generating scene asset previews..." and tell user to wait.
* Only render previews for assets that are currently visible in filesystem dock's grid view, but a search prompt might still freeze the engine for a while if the results have scene files.

Since re-creating thumbnails for the whole project with 100+ scene files should not occur frequently, I guess I'll just slap a progress bar if the thumbnail creation queue is over 10 or more.

![Screenshot from 2025-04-16 05-01-28](https://github.com/user-attachments/assets/0eecf085-af03-492c-bfee-238039e5d844)

KoBeWi commented:
You can avoid blocking the editor completely by waiting a frame between each preview. Though it will make the generation slower and the editor might become laggy.

daniel080400 commented:
> When using Forward/ Mobile renderer (Multi threaded thumbnail creation), the editor stays responsive, but the viewport flickers (to a point it's irritating) while generating thumbnails

Weird, after testing, I found out this is not widely reproducible. Viewport flickering only happens when using Mobile renderer and if the current opened scene has a visible LightMapGI (Flickering does not occur if lightmap is hidden).

Things are just normal when I opened other levels without baked lightmaps.

https://github.com/user-attachments/assets/4f7a704b-83b8-42e0-a1df-9ecff3be5d6c

KoBeWi commented on doc/classes/EditorInterface.xml:
```diff
@@ -422,6 +422,7 @@
 			<param index="1" name="with_preview" type="bool" default="true" />
 			<description>
 				Saves the currently active scene as a file at [param path].
+				[b]Note:[/b] The [param with_preview] parameter has no effect.
```
Do we have a way to deprecate parameters? 🤔

daniel080400 replied:
As far as I know no, this method is exposed, so changing it breaks compatibility.

KoBeWi commented on editor/import/3d/resource_importer_scene.cpp:
```diff
@@ -3266,7 +3266,7 @@ Error ResourceImporterScene::import(ResourceUID::ID p_source_id, const String &p
 		err = ResourceSaver::save(packer, p_save_path + ".scn", flags); //do not take over, let the changed files reload themselves
 		ERR_FAIL_COND_V_MSG(err != OK, err, "Cannot save scene to file '" + p_save_path + ".scn'.");
-		_generate_editor_preview_for_scene(p_source_file, scene);
+		// _generate_editor_preview_for_scene(p_source_file, scene);
```
?
Remove that method if it's unused.

KoBeWi commented on editor/plugins/editor_preview_plugins.h:
```diff
@@ -32,6 +32,8 @@
 
 #include "editor/editor_resource_preview.h"
 
+#define MATH_PI 3.14159265358979323846f
```
Pretty sure you can get that from Math.

daniel080400 replied:
Now uses Math::PI

KoBeWi commented on editor/plugins/editor_preview_plugins.h:
```diff
@@ -66,10 +68,29 @@ class EditorBitmapPreviewPlugin : public EditorResourcePreviewGenerator {
 	virtual Ref<Texture2D> generate(const Ref<Resource> &p_from, const Size2 &p_size, Dictionary &p_metadata) const override;
 	virtual Ref<Texture2D> generate_from_path(const String &p_path, const Size2 &p_size, Dictionary &p_metadata) const override;
+
+	EditorPackedScenePreviewPlugin();
+	~EditorPackedScenePreviewPlugin();
```
What's this for?

daniel080400 replied:
Cleaned up now

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -338,6 +1010,9 @@ Ref<Texture2D> EditorMaterialPreviewPlugin::generate(const Ref<Resource> &p_from
 
 		draw_requester.request_and_wait(viewport);
+		if (EditorResourcePreview::get_singleton()->is_threaded()) {
+			draw_requester.request_and_wait(viewport); // HACK - Prevents incorrect thumbnail assignment when using Forward or Mobile renderer, comment out this line to see the bug.
```
So waiting twice fixes a bug?

daniel080400 replied:
Yes but I still don't know why, root of problem might be inside the renderer. Material previews can be incorrect sometimes even in 4.4.1 master, I experienced it before.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
-	String cache_base = ProjectSettings::get_singleton()->globalize_path(p_path).md5_text();
-	cache_base = temp_path.path_join("resthumb-" + cache_base);
+	// Safe checks, since this function interacts with EditorNode to render previews
+	ERR_FAIL_COND_V_MSG(!Engine::get_singleton()->is_editor_hint(), Ref<Texture2D>(), "This function can only be called from the editor.");
```
The check below implies it already.

daniel080400 replied:
Cleaned up

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+	ERR_FAIL_COND_V(!bundle.has("conns"), false);
+
+	const uint8_t supported_version = 3;
+	uint8_t current_version = 1;
+	if (bundle.has("version")) {
+		current_version = bundle["version"];
+	}
```
```suggestion
	uint8_t current_version = bundle.get("version", 1);
```

daniel080400 replied:
Cleaned up

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+	}
+
+	// Find and remove all scripts in scene
+	Ref<Script> const dummy = 0;
```
```suggestion
	const Ref<Script> dummy;
```

daniel080400 replied:
Cleaned up

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+
+	// Recursively apply to all child scenes
+	for (int i = 0; i < edited_variants.size(); i++) {
+		if (Object::cast_to<PackedScene>(edited_variants[i])) {
+			_remove_scripts_from_packed_scene(edited_variants[i]);
+		}
```
You can move this inside the loop above.

daniel080400 replied:
Done.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+
+	// Create a new scene state
+	bundle["variants"] = edited_variants;
+	Ref<SceneState> new_state = memnew(SceneState);
```
```suggestion
	Ref<SceneState> new_state;
	new_state.instantiate();
```

daniel080400 replied:
Cleaned up

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+
+		// Preview camera
+		Camera2D *camera = memnew(Camera2D);
+		camera->set_name("ThumbnailCamera2D");
```
Why do you set name if it's not used?

daniel080400 replied:
No reason. Cleaned up.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+		preview_root->add_child(camera);
+
+		// Attach subviewport deferred (thread safe)
+		EditorNode::get_singleton()->call_deferred("add_child", sub_viewport);
```
```suggestion
		callable_mp((Node *)EditorNode::get_singleton(), &Node::add_child).call_deferred(sub_viewport);
```

daniel080400 replied:
All deferred calls uses callable_mp() now

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+		}
+
+		// Make 2D camera current
+		camera->make_current();
```
Not needed. Camera becomes active automatically.

daniel080400 replied:
Cleaned up.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+
+		// Retrieve thumbnail of 2D (No GUI)
+		Ref<ImageTexture> capture_2d = ImageTexture::create_from_image(sub_viewport->get_texture()->get_image());
+		if (capture_2d->get_image()->get_size() != p_size) {
```
The check is not needed, `resize()` already checks it.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+		thumbnail->set_image(thumbnail_image);
+
+		// Clean up
+		EditorNode::get_singleton()->call_deferred("remove_child", sub_viewport);
+		EditorNode::get_singleton()->call_deferred("remove_child", sub_viewport_gui);
```
You don't need to remove child before freeing.

daniel080400 replied:
removed.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+		// Clean up
+		EditorNode::get_singleton()->call_deferred("remove_child", sub_viewport);
+		EditorNode::get_singleton()->call_deferred("remove_child", sub_viewport_gui);
+		sub_viewport->call_deferred("queue_free");
+		sub_viewport_gui->call_deferred("queue_free");
```
`queue_free()` already works like deferred method.

daniel080400 replied:
all queue_free() is called without deferred now.

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+		EditorNode::get_singleton()->call_deferred("remove_child", sub_viewport_gui);
+		sub_viewport->call_deferred("queue_free");
+		sub_viewport_gui->call_deferred("queue_free");
+		p_scene->queue_free();
```
`p_scene` is part of viewport, so it doesn't need to be freed separately.

daniel080400 replied:
now only frees subviewports

KoBeWi commented on editor/plugins/editor_preview_plugins.cpp:
```diff
@@ -292,30 +314,680 @@ Ref<Texture2D> EditorPackedScenePreviewPlugin::generate(const Ref<Resource> &p_f
+
+	Rect2 n2d_rect = Rect2(); // The rect of the current iterating node2d
+
+	if (p_node->is_class("Sprite2D")) {
+		Sprite2D *sprite = Object::cast_to<Sprite2D>(p_node);
```
```suggestion
	Sprite2D *sprite = Object::cast_to<Sprite2D>(p_node);
	if (sprite) {
```
Same below.

daniel080400 replied:
Now all uses class casting

KoBeWi commented:
Left some comments for the code, but functionality-wise it's fine. I tested with a big 2D project and it generates nice previews. I didn't test 3D-heavy project and other renderers than Forward+. Needs more testing probably.

Note that invalidating cache will cause heavy CPU and disk load when the new previews are generated for the first time, but it's necessary.

daniel080400 commented:
Changes made:
1. Rewrite most code for 3d thumbnail creation, now uses a SubViewport node under EditorNode to render thumbnails (Instead of interacting with RenderingServer directly), this should be less prone to bug since all references are auto-freed when removed from EditorNode.
2. Done most clean up as @KoBeWi mentioned
3. Fixed error when scene has ```@export``` variables in scene (Those data are written inside .tscn and will spam error if we remove the scripts while generating previews, now fixed.)
4. Renamed ```_remove_scripts_from_packed_scene``` to ```_setup_packed_scene``` as it does more than removing scripts  in scene now
5. Implemented a progress bar if generating thumbnails using **gl_compatiblity** renderer (Single-threaded). Delaying a frame between creating thumbnails is not an option, the editor will be next to unresponsive (only 1-2fps), so we wait.
![Screenshot from 2025-04-21 04-31-34](https://github.com/user-attachments/assets/2eadf21e-c321-448f-ae47-512ba40352fc)
6. Make sure that we only generate thumbnails for a scene at a time, previously there are too many callbacks involved while generating previews, which makes ```EditorResourcePreview::_idle_callback()```and ```EditorResourcePreview::_iterate()``` to be called multiple times while a scene thumbnail is still generating. This is still experimental, I'm still finding a way to deal with large/heavy scenes. Most scenes generates its thumbnail within 1 seconds, but for a large scene it can be over 10 seconds. I would say we should give up creating thumbnails for scene that takes more than ~2 seconds and proceed to next scene, unless it's the current opened scene and just been saved. Opened for discussion.
```
//editor_resource_preview.cpp::440

	// Ensure to process one item at a time
	singleton->preview_mutex.lock();
	bool processing = singleton->processing_item.path != "";
	int last_msec = singleton->last_process_msec;
	singleton->preview_mutex.unlock();
	if (processing) {
		// Took more than 2 seconds to generate thumbnail, might be stuck or bugged, proceed to next item.
		if (OS::get_singleton()->get_ticks_msec() - last_msec > 2000) {
			singleton->preview_mutex.lock();
			singleton->processing_item = QueueItem();
			singleton->preview_mutex.unlock();
		}
		return;
	}
```

Issues:
The viewport flickering issue in Mobile Renderer I previously mentioned, turns out it is very likely a renderer bug, it happens when:
1. Renderer is set to Mobile
2. Current opened scene has LightMapGI baked
3. There's a subviewport somewhere (Even outside of MainTree), with a own World3D and a active Camera3D as child

Then the opened scene's LightMapGI will not render. I tried to find ways to workaround but no luck, this should be fixed directly in Mobile renderer.

KoBeWi commented:
Here's a random test project with lots of junk files:
[TestProject.zip](https://github.com/user-attachments/files/19937540/TestProject.zip)
I can still see randomly appearing windows and DEV_ASSERT spam.
Also the project crashes for some reason when generating previews.

daniel080400 commented:
Changes made:


1. Make EditorNode::add_io_warning() able to be called on thread:
```c++
// editor/editor_node.cpp:7139

callable_mp_static(EditorNode::add_io_warning).call_deferred(p_warning);
```
Previously when creating thumbnails outside of main thread, if the scene instantiates with warning, will trigger ```EditorNode:4788:DEVASSERT(Thread::get_caller_id() == Thread::get_main_id())``` and crash.

2. Abort generating thumbnails for scenes that have **Viewport**, **OptionButton** derived class as root, since it'll lead to crashes:
```c++
// editor/plugins/editor_preview_plugins.cpp:344

// Prohibit Viewport class as root when generating thumbnails
bool prohibited_root_type = (
	Object::cast_to<Viewport>(p_scene) || 
	Object::cast_to<OptionButton>(p_scene) // NOTE - OptionButton uses PopupMenu(Derived class of Viewport) internally
);

if (prohibited_root_type) {
	p_scene->queue_free();
	return Ref<Texture2D>();
}
```

3. Do not skip large scene files, instead display file names in progress bar, so users can have clue which file is slowing down:
![image](https://github.com/user-attachments/assets/53703fbf-1d13-4fa1-bae5-5eebde7fbf7d)


4. Fixed **EditorMeshPreviewPlugin**  not generating previews correctly when running on thread by calling ```draw_requester.request_and_wait()``` twice

5. Do not account nodes that are not visible into scene aabb/rect calculation.

6. Tested 2d scene thumbnails, making preview 2d camera as current is necessary if the scene already has another camera, added relevant code back:
```c++
// editor/plugins/editor_preview_plugins.cpp:475

if (EditorResourcePreview::get_singleton()->is_threaded()) {
	callable_mp((Node *)EditorNode::get_singleton(), &Node::add_child).call_deferred(sub_viewport, false, Node::InternalMode::INTERNAL_MODE_DISABLED);
	callable_mp((Camera2D *)camera, &Camera2D::make_current).call_deferred();
} else {
	EditorNode::get_singleton()->add_child(sub_viewport);
	camera->make_current();
}
```

7. Set all **Window** in scene to invisible, avoid popups.
```c++
// editor/plugins/editor_preview_plugins.cpp:584 && 623

// Don't let window to popup
Window *window = Object::cast_to<Window>(p_node);
if (window) {
	window->set_visible(false);
}
```

---

Manual tests done:
1. Tested various projects including the junk file project @KoBeWi provided, with Mobile, Compatibility Renderer, with no crashes (It's already flooded with loading errors but I see no errors related to thumbnail creation)

2. Can hit play game, exit editor, or edit&save scene while preview is still generating without errors. (Mobile Renderer) Need more extensive tests though, if it eventually run into problem, I'll consider run all thumbnail creation single-threaded for stability.

KoBeWi commented:
> Abort generating thumbnails for scenes that have Viewport, OptionButton derived class as root, since it'll lead to crashes:

What about other classes with popups? MenuButton, ColorPickerButton etc.

> (The junk file project is already flooded with loading errors but I see no errors related to thumbnail creation)

Well, the errors are a result of thumbnail generation, but I guess there is no way around that (other than not updating thumbnails 🙃). Although it's not something that will happen in a regular project without random broken files.

KoBeWi reviewed (approved):
Functionality-wise it looks good now.
The implementation seems fine in general, but I did not review too deeply. There is some usage of mutexes and rendering API that might need extra checking.

daniel080400 commented:
Changes made since b93e5bff:

Revert the behavior of prohibiting **OptionButton** node as root for the preview scene. See explanation below. (Still prohibits **Viewport** class due to rendering issues)
```c++
// editor/plugins/editor_preview_plugins.cpp:344

	// Prohibit Viewport class as root when generating thumbnails
	if (Object::cast_to<Viewport>(p_scene)) {
		p_scene->queue_free();
		return Ref<Texture2D>();
	}
```


---

@KoBeWi 
> What about other classes with popups? MenuButton, ColorPickerButton etc. EDIT: Actually, if Viewport root is a problem, why OptionButton is also a problem if its popup is not root?

After testing, I confirmed that the crash occurs when there's visible **Window** as scene root and attached to **EditorNode**, but I already added code to set all **Window** to invisible in preparation. So I removed the restriction of **OptionButton** roots and tested it again, the project didn't crash.

> > (The junk file project is already flooded with loading errors but I see no errors related to thumbnail creation)
> 
> Well, the errors are a result of thumbnail generation, but I guess there is no way around that (other than not updating thumbnails 🙃). Although it's not something that will happen in a regular project without random broken files.

Correct, the error log is inevitable since we need to load the .tscn to generate thumbnails for it. Unless we can have a method in **ResourceLoader** to load scene but don't print out any loading errors.

daniel080400 commented:
rebased to 99f5a3d665

Calinou reviewed (approved):
Tested locally, it works as expected. Code looks good to me.

### Platformer

![Screenshot_20250530_225506](https://github.com/user-attachments/assets/6445a7fe-4d9e-40db-9bc4-79daa2529041)

### Truck Town

*`town_scene.tscn` has 2D elements visible, but it's primarily a 3D scene. Its root node is a Node3D, so maybe automatic detection can be improved here.*

![image](https://github.com/user-attachments/assets/299355da-49bb-48aa-8cfb-37eeaa68c948)

daniel080400 committed:
Rework scene preview thumbnails

daniel080400 commented:
Changes made since 0834318:
1. Make sure 3d scene thumbnails uses the preview camera respectively
2. Make sure 3d scene thumbnails does not render any Node2D or Control nodes (Might occupy large proportion of the thumbnail)
3. Now uses orthogonal mode for large 3d scenes that exceeds the default camera far (4096 units now)
4. Disable skinning for skeleton meshes, because animation / skeleton scaling might mess around with scene aabb calculation
5. Use ```is_visble_in_tree()``` instead of ```is_visible()``` to determine if the node2d/3d needs to participate scene aabb/rect calculation

---

# Tests
## Platformer 3D
The ```player.tscn``` thumbnail is zoomed out, not because of skeleton, but is affected by the label3D (CoinCount) 

![image](https://github.com/user-attachments/assets/e7cc71f6-e35d-43e1-8f71-45a86c3b86d0)

## Platformer 2D
![image](https://github.com/user-attachments/assets/bed904cc-b082-405c-87e2-196b5393f318)

## Truck Town
The ```town_scene.tscn``` has a very large plane and the town is sitting in the middle of it, it's normal to see almost nothing in the thumbnail since the scene's aabb is very large.

![image](https://github.com/user-attachments/assets/e1c0af2d-bca5-4c3f-9288-6bc97c6e2ac7)

Karagra commented:
> > This should only affect first launch, the previews are not generated if scene has no changes.
> 
> I'll try to investigate later why this doesn't work on MacOS

Thumbnails are generating each time I open my project on Linux also.

jerryjrowe commented:
Also reaffirming that on Linux (Ubuntu), it generates thumbnails on every single project load, and takes a few minutes.

Meorge commented:
On macOS, thumbnails for what seem like *all* assets (including .png, .ogg, .txt, and .blend files - not just scenes) have their thumbnails regenerated whenever I open my project up:

https://github.com/user-attachments/assets/2af73e83-55e9-47a4-a9e4-b4b36bc91011

(The broken dependency issue there shouldn't be relevant, that was just me deleting an unused asset)

I'd be happy to investigate this further to determine what's causing all of the thumbnails to be regenerated if no one else happens to figure it out first, but I may not be able to get to it for a few days.

KoBeWi commented:
See if #107719 fixes your issue.

Meorge commented:
Just tested, and unfortunately it doesn't - I still see the editor generate thumbnails for lots of non-scene assets.

KoBeWi commented:
Please open a new issue.

ChronoDK commented:
Is there a way to disable this feature? Not enjoying the time it adds to load time, and most of my thumbnails ends up looking kinda weird/useless anyway (maybe my scenes are badly setup, but still... would like to just disable this feature).

akien-mga commented:
On second thought after discussing it with other maintainers - there are some aspects to this implementation which I missed initially when approving it to merge, and which aren't strictly speaking bugs, but design problems.

Godot's previous behavior for scene thumbnails is to generate a preview *when saving that scene*. So there's a minor slowdown on save but not too bothersome. It means that some scenes don't have a preview at first, and only get one after they're edited and saved.

This PR seems to change it to load all scenes on load to generate previews, which can be very slow and a hard tradeoff for big projects for what's mostly a cosmetic feature (which a lot of users don't even use as they use the filesystem dock in List view, which is the default - though I assume it's also helpful in the EditorFileDialog and QuickOpen dialog).

This loading time cost is paid only once in theory (if we can fix #107736), but it brings other potential worse problems: Loading all scenes in a thread can lead to errors or deadlocks due to cyclic references or waiting on the renderer. This may not necessarily be the case here, nor a direct consequence of this PR, but it's something we've been struggling with a lot in the past with the preview generation, and so I'm wary of changing it so drastically at this stage of development. We should definitely watch out for reports saying they can't open their projects in 4.5 beta 1 anymore, which could be traced back to this.

So I think we should consider the following changes:

- Don't generate scene thumbnails on load, and keep the current system where thumbnails are generated on scene save.
- Add an option (under Project > Tools) to scan and generate all thumbnails for the project, for users who want it. I.e. change this PR's "on load" behavior to one that's on demand.
- Benchmark before/after preview generation on bigger scenes to ensure we're not making it unreasonably slow. See e.g. discussion on MSAA 8x in #107689.

Given the phase we are in with regard to the 4.5 release, it may be reasonable to revert this change for 4.5 beta 2 and redo it for 4.6 with the above changes and reported bugs addressed. Merging it again early in the 4.6 dev cycle would then give us more time to properly address any remaining issue.

JoNax97 commented:
> Don't generate scene thumbnails on load, and keep the current system where thumbnails are generated on scene save.
Add an option (under Project > Tools) to scan and generate all thumbnails for the project, for users who want it. I.e. change this PR's "on load" behavior to one that's on demand.

What about generating previews on demand when the editor actually needs to show them? 

Yes, the thumbnails will pop in the UI which is a bit ugly, but I think it solves the bottleneck of generating them for all scenes in the project, while being more elegant than a manual generate button.

Jordyfel commented:
After the revert I noticed preview images from this PR did not have rounded corners, while the previous (now restored) system has them

daniel080400 commented:
I just got time recently, now I'm testing the new code ( This PR + Bug Fixes ), should be stable enough to open a new PR this week.

daniel080400 commented:
Please continue discussion at  #112060
